text/2018-11-12-backup-restore-replication-with-raft-learner.md

```diff
@@ -0,0 +1,168 @@
+                                     |          |      
+                                     |          |      
++---------+                          | +---+---+|      
+|Region 2 |                          +>|1  |2  ||      
++---------+                            +---+---+|      
+                                                v      
+```                                                                             
+
+When we do the checkpoint, the checkpoint has already contained Region 1 and Region 2. Region 1 will be replayed from the Snapshot, and Region 2 will be replayed from the first log. But here is a problem when we replay Region 2 before Region 1. 
+
+Assume the command in the log 1 of Region 2 is `SET b = 10`, but the command in the log 11 of Region 1 is `SET b = 9`. So if Region 2 is replayed first, the `b` will have a wrong value. To solve this problem, we need to build a Replay order, the implementation may be easy - just to record the event.
+
+For Split 1 -> 2 + 1, Region 2 must record a event that `Region 2 is split from Region 1 at index 100`. When Region 2 wants to replay, it finds that - “oh, Region 1 hasn’t applied to 100 yet, so let’s wait”.
+
+For merge 1 + 2 -> 1, Region 1 must also persist the similar event, and it must wait for Region 2's replaying and then starts to replay later.
+
+Using this way has another benefit, now we must sync the state machine forcibly when we meet Split, Merge, or other Admin commands because we have no way to recover the cluster with only replaying Raft logs. Sync can cause latency increase sometimes, so removing this can help us increase the performance.
+
+So for restore, we need to first build the Replay tree (of course, there must be some problems when we start to do this) and replay the Regions concurrently or in order. This may be similar to lightning. We can create the Region directly in the new cluster, ingest the snapshot to that Region, and then add the following logs. After doing this, the Region can itself apply the logs and reach a consistency. For better performance, we can let the new leader fetch the snapshot and following Raft logs from Ceph directly. 
+
+There will be another RFC to discuss how to recover the cluster just from Raft logs. 
+
+## Node HA
+
+If the Backup crashed and can’t be recovered, we may lose the backup data. A simple way is to add a new Backup node directly.
+
+Another way is to back up the metadata in Ceph regularly and let the new Backup node resume the replication easily. 
+
+# Drawbacks
+
+As you can see, there is one Backup node in the design, this may become the bottleneck. The snapshot transportation or log replication may occupy the whole network of this node. Another problem is whether the single node can handle too many Regions.
+
+To solve this problem, maybe we can use multi Backup nodes, each node is responsible for multi-region. For example, if TiDB is running on TiKV, we can use different Backup nodes for different databases.
```

Comment:
Does it mean we stop backup a database if its backup node fails? Why not let PD to schedule regions to different nodes to avoid a single point of failure.